### PR TITLE
Survey ID should be unique per study

### PR DESCRIPTION
**Pull request by DwayneJengSage**
https://sagebionetworks.jira.com/browse/BRIDGE-1833

Testing done: unit tests and integ tests

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDao.java [hunk 1]
```diff
@@ -378,7 +377,35 @@ public Survey getSurvey(GuidCreatedOnVersionHolder keys, boolean includeElements
...
+        // Range key.
+        Condition rangeKeyCondition = new Condition().withComparisonOperator(ComparisonOperator.EQ)
+                .withAttributeValueList(new AttributeValue().withS(surveyId));
+
+        // Construct query.
+        DynamoDBQueryExpression<DynamoSurvey> expression = new DynamoDBQueryExpression<DynamoSurvey>()
+                .withConsistentRead(false)
+                .withHashKeyValues(hashKey)
+                .withRangeKeyCondition("identifier", rangeKeyCondition)
+                .withLimit(1);
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDao.java:397]: Don't we need to use the Index object if querying using the index you've specified for studyId-identifier?

File: app/org/sagebionetworks/bridge/services/SurveyService.java [hunk 2]
```diff
@@ -99,6 +104,16 @@ public Survey createSurvey(Survey survey) {
...
+
+        // Check survey ID uniqueness.
+        StudyIdentifier studyId = new StudyIdentifierImpl(survey.getStudyIdentifier());
+        String existingSurveyGuid = surveyDao.getSurveyGuidForIdentifier(studyId, survey.getIdentifier());
+        if (existingSurveyGuid != null) {
+            String errMsg = "Survey identifier " + survey.getIdentifier() + " is already used by survey " +
+                    existingSurveyGuid;
+            Map<String, Object> entityKeyMap = ImmutableMap.of(KEY_IDENTIFIER, survey.getIdentifier());
+            throw new EntityAlreadyExistsException(Survey.class, errMsg, entityKeyMap);
+        }
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/SurveyService.java]: To my mind, this kind of a constraint would be thrown before we'd throw a validation exception.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDaoTest.java [hunk 3]
```diff
@@ -54,21 +54,23 @@
 @ContextConfiguration("classpath:test-context.xml")
 @RunWith(SpringJUnit4ClassRunner.class)
 public class DynamoSurveyDaoTest {
-
-    private static Logger logger = LoggerFactory.getLogger(DynamoSurveyDaoTest.class);
+    private static final Logger logger = LoggerFactory.getLogger(DynamoSurveyDaoTest.class);
```
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDaoTest.java]: by convention everywhere else we've leaned on calling this LOG I think (or maybe LOGGER). It's fine with me if we standardize on this as we came across it.

**Review by alxdarksage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDao.java (hunk 1 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDao.java:397]: Apparently not, if integration tests are any indication.

File: app/org/sagebionetworks/bridge/services/SurveyService.java (hunk 2 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/services/SurveyService.java]: Moved this to before the validation check.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDaoTest.java (hunk 3 above)
Comment (DwayneJengSage) [reply test/org/sagebionetworks/bridge/dynamodb/DynamoSurveyDaoTest.java]: Renamed to LOG.

**Review by alxdarksage — APPROVED** [review]

**Pull request merged**